arraysandpointers.tex

```diff
@@ -1,144 +1,164 @@
 
-A pointer is declared by adding a * between the type and the variable name in a normal declaration.
-One thing you must be careful of is the associativity of the *.
-The * binds to the variable \emph{name}, not the type!
-This is a common way people get tricked on exams.
-A pointer can also point to another pointer.
+We can find the location of the first element in an array with \lstinline!&array[0]!.
+Indexing into \lstinline!array! gives us the value of the first element, and the address-of operator \lstinline!&! gives us its memory address.
+If we wanted to find the address of the second element we would use \lstinline!&array[1]!.
+Unfortunately, this doesn't help us if we don't have access to the \lstinline!array! variable.
+Given that the elements of an array are laid out in a contiguous line, and are all the same size\footnote{An item of a specific type has a fixed size in C}, there should be a way to calculate the address of the second element from the address of the first.
 
 \begin{codeblock}
-int *p; // Pointer to int
-int** p1; // Pointer to a pointer to an int
-int* x, y; // x is pointer to int, y is int!
+int array[5];
+int* firstElement = &array[0];
+int* secondElement = firstElement + 1; // equivalent to &array[1]
 \end{codeblock}
 
-\subsection{Getting the address of something}
+This is known as \emph{pointer arithmetic}.
+The value of a pointer is a numerical location, so by adding to it we get a location further along in memory.
+Notice that we wrote \lstinline!firstElement + 1! above; while memory addresses are normally thought of in bytes, C realises that \lstinline{firstElement} is a pointer to an \lstinline!int! and gives us a pointer 1 \lstinline!int! along from it, rather than 1 byte along.
+When using pointer arithmetic, C will always figure out how many bytes it should add to or subtract from the pointer to move the correct number of elements along.
 
-There are two ways you can get the address of something.
-The first is to steal it from somebody else.
-If you already have a pointer variable, you can assign it to another variable just like you would any other type.
+This is a very convenient pattern, and it turns out to not be a mere coincidence.
+Indexing an array and performing pointer arithmetic are the exact same thing!
+Remember that an array is really just a block of memory --- there is no better way to refer to it than by the memory address it starts at, and this is what C does.
+When we declare an array \lstinline!T array[n]!, the variable \lstinline!array! can be used as a pointer to the start of the array's memory, which is also the address of its first element.
+This means that where we were previously using \lstinline!&array[0]! we can now just use \lstinline!array!.
 
 \begin{codeblock}
-int* p = doSomething(); // Returns an int *
-int* p1 = p; // p1 points to the same thing p does.
+/* These are both the same */
+*(array + 1) = 1;
+array[1] = 1; 
```

Comment:
Perhaps mention that square brackets are the more common and preferred method for this course?